frame/staking/src/benchmarking.rs

```diff
@@ -0,0 +1,651 @@
+		.collect();
+
+
+	// compact encode the assignment.
+	let compact = <CompactOf<T>>::from_assignment(
+		low_accuracy_assignment,
+		nominator_index,
+		validator_index,
+	).unwrap();
+
+	// winners to index.
+	let winners = winners.into_iter().map(|w|
+		snapshot_validators.iter().position(|v| *v == w).unwrap().try_into().unwrap()
+	).collect::<Vec<ValidatorIndex>>();
+
+	(winners, compact, score)
+}
+
+fn get_seq_phragmen_solution<T: Trait>(do_reduce: bool)
+-> (Vec<ValidatorIndex>, CompactOf<T>, PhragmenScore) {
+	let sp_phragmen::PhragmenResult {
+		winners,
+		assignments,
+	} = <Module<T>>::do_phragmen::<Percent>().unwrap();
+
+	let snapshot_validators = <Module<T>>::snapshot_validators().unwrap();
+	let snapshot_nominators = <Module<T>>::snapshot_nominators().unwrap();
+
+	offchain_election::prepare_submission::<T>(
+		snapshot_nominators,
+		snapshot_validators,
+		assignments,
+		winners,
+		do_reduce,
+	)
+}
+
+#[cfg(test)]
+fn clean<T: Trait>() {
+	use frame_support::StoragePrefixedMap;
+	<Validators<T>>::enumerate().for_each(|(k, _)| {
+		let ctrl = <Module<T>>::bonded(&k).unwrap();
+		<Bonded<T>>::remove(&k);
+		<Validators<T>>::remove(&k);
+		<Stakers<T>>::remove(&k);
+		<Ledger<T>>::remove(&ctrl);
+	});
+	<Nominators<T>>::enumerate().for_each(|(k, _)| <Nominators<T>>::remove(k));
+	<Stakers<T>>::remove_all();
+	<Ledger<T>>::remove_all();
+	<Bonded<T>>::remove_all();
+	<QueuedElected<T>>::kill();
+	QueuedScore::kill();
+}
+
+#[repr(u32)]
+#[allow(dead_code)]
+#[derive(Debug)]
+pub enum BenchmarkingMode {
+	/// Initial submission. This will be rather cheap
+	InitialSubmission,
+	/// A better submission that will replace the previous ones. This is the most expensive.
+	StrongerSubmission,
+	/// A weak submission that will be rejected. This will be rather cheap.
+	WeakerSubmission,
+}
+
+/// Benchmark `set` extrinsic.
+struct SubmitElectionSolution;
+impl<T: Trait> BenchmarkingSetup<T, Call<T>, RawOrigin<T::AccountId>> for SubmitElectionSolution
+	where T::Lookup: StaticLookup<Source=AddressOf<T>>
+{
+	fn components(&self) -> Vec<(BenchmarkParameter, u32, u32)> {
+		vec![
+			// number of nominators
+			(BenchmarkParameter::N, 1, 5_000),
+			// number of validator candidates
+			(BenchmarkParameter::V, 1, 1000),
+			// num to elect
+			(BenchmarkParameter::E, 1, 500),
+			// edge per vote
+			(BenchmarkParameter::Z, 2, 16),
+		]
+	}
+
+	fn instance(&self, components: &[(BenchmarkParameter, u32)])
+		-> Result<(Call<T>, RawOrigin<T::AccountId>), &'static str>
+	{
+		let num_validators = components.iter().find(|&c| c.0 == BenchmarkParameter::V).unwrap().1;
+		let num_nominators = components.iter().find(|&c| c.0 == BenchmarkParameter::N).unwrap().1;
+		let edge_per_voter = components.iter().find(|&c| c.0 == BenchmarkParameter::Z).unwrap().1;
+		let mode: BenchmarkingMode = BenchmarkingMode::WeakerSubmission;
+		let do_reduce: bool = true;
+		let to_elect: u32 = components.iter().find(|&c| c.0 == BenchmarkParameter::E).unwrap().1;
+
```

Comment:
Mmm, I wonder if we should start building on top of the branch of Gav, it seems almost finished.

Reply:
Regardless of the benchmarking details, this also has one or two code updates that are needed in the base branch. So I'd prefer getting this stuff asap into the base.